lib/vtls/openssl.c

```diff
@@ -3027,214 +3027,218 @@ static CURLcode populate_x509_store(struct Curl_cfilter *cf,
   if(!store)
     return CURLE_OUT_OF_MEMORY;
 
-  if(verifypeer) {
 #if defined(USE_WIN32_CRYPTO)
-    /* Import certificates from the Windows root certificate store if
-       requested.
-       https://stackoverflow.com/questions/9507184/
-       https://github.com/d3x0r/SACK/blob/master/src/netlib/ssl_layer.c#L1037
-       https://datatracker.ietf.org/doc/html/rfc5280 */
-    if(ssl_config->native_ca_store) {
-      HCERTSTORE hStore = CertOpenSystemStore(0, TEXT("ROOT"));
-
-      if(hStore) {
-        PCCERT_CONTEXT pContext = NULL;
-        /* The array of enhanced key usage OIDs will vary per certificate and
-           is declared outside of the loop so that rather than malloc/free each
-           iteration we can grow it with realloc, when necessary. */
-        CERT_ENHKEY_USAGE *enhkey_usage = NULL;
-        DWORD enhkey_usage_size = 0;
-
-        /* This loop makes a best effort to import all valid certificates from
-           the MS root store. If a certificate cannot be imported it is
-           skipped. 'result' is used to store only hard-fail conditions (such
-           as out of memory) that cause an early break. */
-        result = CURLE_OK;
-        for(;;) {
-          X509 *x509;
-          FILETIME now;
-          BYTE key_usage[2];
-          DWORD req_size;
-          const unsigned char *encoded_cert;
+  /* Import certificates from the Windows root certificate store if requested.
```

Comment:
You report a problem on Ubuntu, but yet the largest chunk of your PR changes code that is Windows-specific. It makes me assume it changes more than you intended?

Reply:
That large chunk (from my commit 27675b4e3f86e050efb352157edea659b976dedf) is a `git revert` of a3bcfab4b577dde2ddac0f25bbe872037dd81d23 which was the easiest way I could see to remove the `if(verifypeer){` statement starting on [line 3079](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3079) that introduced the dependency between `CURLOPT_SSL_VERIFYPEER` and `CURLOPT_CAINFO_BLOB` options. There should be no functional change to the Windows certificate handling code, though my change would also revert a [small formatting change on line 3006](https://github.com/curl/curl/pull/11457/commits/27675b4e3f86e050efb352157edea659b976dedf#diff-03eaf320162ec20f67fab4e16f494e29735dadfa12425530869d42eea86bd2ccR3006).

Reply:
Though I didn't consider curl's behavior on Windows if `CURLOPT_SSL_VERIFYPEER` was set to 0 and a failure occurred while curl was importing certificates from Windows.

Reply:
Since you removed the `if(verifypeer) {` for the Windows code that loads native certs as well as for the code you want to change, it certainly changes behavior in a non-intended way.

Reply:
After examining the code I didn't find that my changes would have any unintended affect on the Windows native certificate import code.

a3bcfab4b5 is the latest commit that changed the Windows native certificate import code. This means there haven't been any changes since that commit that would need to be preserved. See the code block at the bottom of this comment for my analysis commands.

Before commit a3bcfab4b577dde2ddac0f25bbe872037dd81d23 there were 6 if statements contained in the [populate_x509_store](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3058) function ([1](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3084-L3085), [2](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3223), [3](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3240), [4](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3284), [5](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3292), [6](https://github.com/curl/curl/blob/72d4f7d8f7733b9bfb501f34c18d513c238714cd/lib/vtls/openssl.c#L3309)). The a3bcfab4b577dde2ddac0f25bbe872037dd81d23 commit moved the first 2 if statements into the `if(verifypeer)` statement from [line 3079](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3079) ([1](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3086), [2](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3223)). The remaining if statements were not changed ([3](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3235), [4](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3279-L3280), [5](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3288), [6](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3316)). However moving these if statements only affected the behavior of [the second](https://github.com/curl/curl/blob/a3bcfab4b577dde2ddac0f25bbe872037dd81d23/lib/vtls/openssl.c#L3223), making it dependent on the configuration of `VERIFYPEER`.

For Windows, before a3bcfab4b577dde2ddac0f25bbe872037dd81d23, certificates would only be loaded from Windows with the `CURLSSLOPT_NATIVE_CA` flag set in `CURLOPT_SSL_OPTIONS` and either `VERIFYPEER` or `VERIFYHOST` was set. After the a3bcfab4b577dde2ddac0f25bbe872037dd81d23 change, certificates from Windows would only be loaded if both the `NATIVE_CA` flag and `VERIFYPEER` were set. I believe the a3bcfab4b577dde2ddac0f25bbe872037dd81d23 commit changes the semantics around the `CURLOPT_SSL_VERIFYHOST` option, and my change would return to the original semantics. After [my changes](https://github.com/kyled-dell/curl/blob/5519a735a5c047f771ba1494ecb54a503b6b0350/lib/vtls/openssl.c#L3035-L3036) the first if statement would exactly return to its curl 7.87.0 behavior. Curl's 7.87.0 Windows native CA behavior is identical to [curl 7.86.0's behavior](https://github.com/curl/curl/blob/cd95ee9f771361acf241629d2fe5507e308082a2/lib/vtls/openssl.c#L3205-L3206), which is the behavior I hope to preserve through this PR. The behavior of Windows' usage of `VERIFYPEER` with and without my change remains the same. With my change, configuring `VERIFYHOST` can also be used to enable certificate import from Windows.

The [comment on line 3047](https://github.com/curl/curl/blob/f2aac0d108167bee5d06d6f9a307e0ff5193b816/lib/vtls/openssl.c#L3048-L3051) states that errors encountered while importing certificates from Windows are ignored. This comment describes the same behavior that my PR is attempting to implement for `CAINFO_BLOB`. However the Windows native code would not execute if `VERIFYPEER` is false, so the error handling behavior would not be executed.

My change will remove the second if statement from the `verifypeer` if statement as it is intends to do. This will restore the independence of the `SSL_VERIFYPEER` and `CAINFO_BLOB` settings. This will preseve functionality that identical to curl 7.86.0.

```text
$ git rev-parse --verify HEAD
f2aac0d108167bee5d06d6f9a307e0ff5193b816
$ git blame lib/vtls/openssl.c -L 3037,3172 | awk '{print $1}' | sort | uniq
3c16697ebd
3d919440c8
49a6642f01
8fa8df95fb
a20daf90e3
a3bcfab4b5
^ae1912cb0
d288222e80
fa1ae0abcd
```